### PR TITLE
Use istio to route traffic, add support for different versions of one…

### LINKED ISSUE #5: Use istio to create canary/blue-green/etc releases

### PR DESCRIPTION
**Pull request by SniXosha**
* Added support for multiple versions of one service 
* New API:
```
osdf deploy component -m replace - deploy component replacing other versions
osdf deploy component -m hidden - deploy component not affecting other versions
osdf route component{v2} 5 - route 5% of traffic to v2 version
osdf route component{v2} mirror - mirror traffic  to version v2
osdf route component{v2} header - add possibility to access v2 version using header
```
* New status example: 

```
COMPONENT                             VERSION    TRAFFIC    STATUS      REPLICAS
job-component                         v1         -          SUCCEEDED   job
example-producer{v4}(not deployed)               
                                      v1         90%,header RUNNING     1/1
                                      v2         10%        RUNNING     1/1
                                      v3         -          RUNNING     1/1
some-other-component{v2}               
                                      v1         90%        RUNNING     1/1
                                      v2         mirror     FAILED      1/1
another-component{v1}                 ?          ?          NOT_FOUND   ?
```

### CODE REVIEW

File: src/main/java/io/microconfig/osdf/istio/rules/MainRule.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    public void setMirror(String subset) {
+        WeightRoute otherRoute = getOtherRoute(subset);
+        routes = of (
+                weightRoute(otherRoute.getDestination(), 100)
+        );
+        mirror = destination(otherRoute.getDestination().getHost(), subset);
+    }
+
+    public void deleteSubset(String subset) {
+        if (mirror != null && mirror.getSubset().equals(subset)) mirror = null;
```
Comment (microconfig) [inline src/main/java/io/microconfig/osdf/istio/rules/MainRule.java:57]: use {}

File: src/main/java/io/microconfig/osdf/istio/rules/MainRule.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        List<WeightRoute> otherRoutes = getOtherRoutes(subset);
+        if (otherRoutes.size() == 1) {
+            routes = of(
+                    weightRoute(otherRoutes.get(0).getDestination(), 100)
+            );
+        } else if (otherRoutes.size() == 0) {
+            routes = of();
```
Comment (microconfig) [inline src/main/java/io/microconfig/osdf/istio/rules/MainRule.java:64]: what if otherRoute.size() > 1 ?

File: src/main/java/io/microconfig/osdf/api/parameter/CanaryWeightParameter.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import static io.microconfig.osdf.utils.StringUtils.castToInteger;
+
+public class CanaryWeightParameter extends AbstractParameter<Integer> {
+    public CanaryWeightParameter() {
+        super("weight", "w", "Traffic weight for canary release. Integer from 0 to 100");
+    }
+
+    @Override
+    public Integer get() {
+        if (getValue() == null) return null;
```
[the remark this answers is not in the document]
Comment (SniXosha) [reply src/main/java/io/microconfig/osdf/api/parameter/CanaryWeightParameter.java:14]: Indication that parameter is not set

File: src/main/java/io/microconfig/osdf/api/OSDFApiImpl.java [hunk 4]
```diff
@@ -106,6 +119,24 @@ public void howToStart() {
...
+                weightRule(oc),
+                mirrorRule(oc),
+                headerRule(oc)
+        );
+    }
+
+    private Deployer deployer(String mode) {
+        if (mode == null || mode.equals("replace")) {
+            return replaceDeployer(oc);
+        } else if (mode.equals("hidden")) {
```
[the remark this answers is not in the document]
Comment (SniXosha) [reply src/main/java/io/microconfig/osdf/api/OSDFApiImpl.java:133]: why?

File: src/main/java/io/microconfig/osdf/commands/StatusCommand.java [hunk 5]
```diff
@@ -42,25 +40,41 @@ public void run(List<String> components) {
...
-    private void addDeploymentComponent(ColumnPrinter printer, DeploymentComponent component, DeploymentInfo info) {
-        String replicasInfo = (info.getStatus() == UNKNOWN || info.getStatus() == NOT_FOUND) ? "?/?" : info.getAvailableReplicas() + "/" + info.getReplicas();
-        printer.addRow(component.getName(), info.getStatus().toString(), replicasInfo, info.getProjectVersion(), info.getConfigVersion());
-        for (int i = 0; i < info.getPods().size(); i++) {
-            printer.addRow(Logger::info, " - " + info.getPods().get(i).getName(), info.getPodsHealth().get(i) ? "OK" : "NOT_READY", "", "", "");
+    private void addComponentHeader(DeploymentComponent component, ColumnPrinter localPrinter, List<DeploymentComponent> deployedComponents) {
+        boolean deployed = deployedComponents.stream().map(DeploymentComponent::getVersion).anyMatch(v -> v.equals(component.getVersion()));
+        String fullName = component.getName() + "{" + component.getVersion() + "}";
+        if (!deployed) {
+            fullName += "(not deployed)";
```
[the remark this answers is not in the document]
Comment (SniXosha) [reply src/main/java/io/microconfig/osdf/commands/StatusCommand.java:76]: it's not reassigning

**Pull request merged**